src/org/opensolaris/opengrok/util/OptionParser.java

```diff
@@ -0,0 +1,701 @@
+ * 
+ * Now that Java 8 has the crucial Lambda and Consumer interfaces, we can
+ * implement a more powerful program option parsing mechanism, ala ruby
+ * OptionParser style.
+ * 
+ * Features
+ *   o  An option can have multiple short names (-x) and multiple long
+ *      names (--xyz). Thus, an option that displays a programs usage
+ *      may be available as -h, -?, --help, --usage, and --about.
+ *   
+ *   o  An option may be specified as having no argument, an optional
+ *      argument, or a required argument. Arguments may be validated
+ *      against a regular expression pattern or a list of valid values.
+ * 
+ *   o  The argument specification and the code to handle it are 
+ *      written in the same place. The argument description may consist
+ *      of one or more lines to be used when displaying usage summary.
+ * 
+ *   o  The option summary is produced without maintaining strings
+ *      in a separate setting.
+ * 
+ *   o  Users are allowed to enter initial substrings for long option
+ *      names as long as there is no ambiguity.
+ * 
+ *   o  Supports the ability to coerce command line arguments into objects.
+ *      This class readily supports Boolean (yes/no,true/false,on/off), 
+ *      Float, Double, Integer, and String[] (strings separated by comma) 
+ *      objects. The programmer may define additional coercions of their own.
+ * 
+ * @author Steven Haehn
+ */
+public class OptionParser {
+
+    // Used to hold data type converters
+    static private HashMap<Class,DataParser> converters = new HashMap<>();
+    
+    static class DataParser {
+        Class dataType;
+        Function<String,Object> converter;
+        
+        public DataParser(Class cls, Function<String,Object> converter) {
+            this.dataType = cls;
+            this.converter = converter;
+        }
+    }
+
+    // Supported internal data type converters.
+    static {
+        accept(Integer.class, s -> { return Integer.parseInt(s); });
+        accept(Boolean.class, s -> { return parseVerity(s); });
+        accept(Float.class,   s -> { return Float.parseFloat(s); });
+        accept(Double.class,  s -> { return Double.parseDouble(s); });
+        accept(String[].class,s -> { return s.split(",");});
+    }
+    
+    // Option object referenced by its name(s)
+    final private HashMap<String,Option> options;
+    
+    // List of options in order of declaration
+    final private List<Option> optionList;
+    
+    // Keeps track of separator elements placed in option summary
+    final private List<Object> usageSummary;
+    
+    private boolean scanning = false;
+    
+    public String prologue;  // text emitted before option summary
+    public String epilogue;  // text emitted after options summary
+    
+    class DuplicateOptionName extends Exception {
+        public DuplicateOptionName(String message) {
+            super(message);
+        }
+    }
+    
+    public class Option {
```

Comment:
general note for this class - all public methods should have a javadoc (maybe besides getters/setters)

Reply:
yes I can do that. Originally, this was not public, but I had to make it so because when I started working in Indexer.java, it couldn't see its methods.